core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1768,7 +1759,9 @@ class UnifiedLog(@volatile var logStartOffset: Long,
       lock synchronized {
         localLog.checkIfMemoryMappedBufferClosed()
         producerExpireCheck.cancel(true)
-        leaderEpochCache.foreach(_.clear())
+        // `renameDir` with `shouldReinitialize=false` sets this to `null` and it's usually (but not always) called before this method
+        if (leaderEpochCache != null)
+          leaderEpochCache.clear()
```

Comment:
@junrao This is interesting - do you think we should be setting the cache to `null` here too versus calling `clear` on it? It's a bit odd that we take different actions to achieve a similar outcome (clearing the in-memory state) in two different (but related) methods.

Reply:
The case where we don't call `renameDir` before `delete` is 

```java
if (logDir.getName.endsWith(UnifiedLog.DeleteDirSuffix)) {
      addLogToBeDeleted(log)
    } 
```